doc/port_fec-ber/port_fec_ber.md

```diff
@@ -0,0 +1,205 @@
+- SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES
+  - monotonically increasing
+  - return not support if its not working for an interface
+
+
+### 4.2 Sai Counters Used
+
+The following redis DB entries will be access for the BER calculations
+
+|Redis DB |Table|Entries|New, RW| Format | Descriptions|   
+|--------------|-------------|------------------|--------|----------------|----------------|  
+|COUNTER_DB |COUNTERS |SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS |R |number |Total number bits corrected</sub>|
+|COUNTER_DB |COUNTERS |SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES |R |number |Tota number uncorrectable frame |
+|COUNTER_DB |COUNTERS_PORT_NAME_MAP |SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES|R |number |Oid to name mapping |  
+|COUNTER_DB |RATES |SAI_PORT_STAT_IF_IN_FEC_CORRECTED_BITS_last | New, RW|number |Last corrected bits counts |
+|COUNTER_DB |RATES |SAI_PORT_STAT_IF_IN_FEC_NOT_CORRECTABLE_FRAMES_last |New, RW|number |Last uncorrctedble frame counts |  
+|COUNTER_DB |RATES |FEC_PRE_BER |New, RW| floating |calculated pre fec BER |  
+|COUNTER_DB |RATES |FEC_POST_BER |New, RW| floating | calulated post fec BER |  
+|APPL_DB |PORT_TABLE |lanes |R |strings |number of serdes lanes in the port |  
+|APPL_DB |PORT_TABLE |speed |R |number |port speed |  
+
+
+### 4.3 SAI API
+
+No change in the SAI API. No new SAI object accessed.
+
+
+### 4.4 Calculation Formulas
+
+Each port can be made up of multiple lanes and each running at the same serdes speed. The hardware counter is per port basis. Therefore the BER calculation will require to account for #lanes and serdes speed of the port.  The lanes count and port speed can be retrieved from the APPL_DB. Serdes speed will be calculated using port speed / number of lanes.
+
+However in the case of post FEC, the frames are actually dropped. There is no way to tell the actual number of error bits in the frame. The calculation is assuming the worst case where all bits in the frame were corrupted.
+
+
+
+```
+Pseudocode :
+
+Step 1: get lanes count and port speed
+    lanes count  : APPL_DB , hget PORT_TABLE lanes , lane count calculate using string.gsub()
+    port speed : APPL_DB , hget PORT_TABLE speed,  speed of the port
+
+Step 2: calculate port_data_rate
+    port_data_rate = math.fmod(port speed / number of lanes)
+
+Step 3 : look up link serdes speed
+    Look up the serdes speed using the following logic
+
+    if port_data_rate == 1000 then
+        serdes = 1.25e+9
+    elseif port_data_rate == 10000 then
+        serdes = 10.3125e+9
+    elseif port_data_rate == 25000 then
+        serdes = 25.78125e+9
+    elseif port_data_rate == 50000 then
+        serdes = 53.125e+9
+    elseif port_data_rate == 100000 then
+        serdes = 106.25e+9
+    else
+        serdes = 0
+    end
+
+Step 4 : frame size
+
+     /* NRZ encoding Rs-528  or PAM4 encoding Rs-544*/
+    if user_port_lane_speed <= 25G
+      frame_size = 5280
+    else
+      frame_size = 5440
+
```

Comment:
Why the post FEC BER errors are multiplied by the frame_size ?  This is the worst case for amount of errors.
For any uncorrectable FEC block, the bit errors can be between 8 to 5280 for <=25G and 16 to 5440 for >=50G.
The lowest possability is the most likely statistically ( 8 or 16)

Reply:
This is a questions of choice.  Since we have no idea the "number of bits" which were bad in this case, we need to take a guess, either the best case as you suggested or the worst case which we were using at this moment.
The way I will suggest how to interpret this value should be  anything other than zero, we have issue.

Reply:
Taking the worst-case scenario (5440) is the most unlikely statistically and might present a false picture.
for a random error distribution, with a typical post fec average BER of 1e-12 it's 1 error for evert 1e12 bits which is ~ 1 error per ~1e8 Fec blocks.

Reply:
look like folks prefer  the statistical average when calculate the post fec,  will make the changes

Reply:
doc updated to use average for post fec BER.